featurebyte/service/feature_table_cache.py

```diff
@@ -1,111 +1,367 @@
     """
 
-    document_class = FeatureTableCacheModel
-
     def __init__(
         self,
-        user: Any,
-        persistent: Persistent,
-        catalog_id: Optional[ObjectId],
-        block_modification_handler: BlockModificationHandler,
-        observation_table_service: ObservationTableService,
+        feature_table_cache_metadata_service: FeatureTableCacheMetadataService,
+        namespace_handler: NamespaceHandler,
+        session_manager_service: SessionManagerService,
+        entity_validation_service: EntityValidationService,
+        tile_cache_service: TileCacheService,
+        feature_list_service: FeatureListService,
     ):
-        super().__init__(
-            user=user,
-            persistent=persistent,
-            catalog_id=catalog_id,
-            block_modification_handler=block_modification_handler,
-        )
-        self.observation_table_service = observation_table_service
+        self.feature_table_cache_metadata_service = feature_table_cache_metadata_service
+        self.namespace_handler = namespace_handler
+        self.session_manager_service = session_manager_service
+        self.entity_validation_service = entity_validation_service
+        self.tile_cache_service = tile_cache_service
+        self.feature_list_service = feature_list_service
 
-    async def get_document_for_observation_table(
+    async def split_cached_and_non_cached_nodes(
         self,
-        observation_table_id: PydanticObjectId,
-    ) -> Optional[FeatureTableCacheModel]:
-        """Get document for observation table.
+        feature_table_cache_metadata: FeatureTableCacheMetadataModel,
+        graph: QueryGraph,
+        nodes: List[Node],
+    ) -> Tuple[Dict[str, Node], Dict[str, Node]]:
+        """
+        Given an observation table, graph and set of nodes
+            - compute nodes definition hashes
+            - lookup existing Feature Table Cache metadata
+            - filter out nodes which are already added to the Feature Table Cache
+            - return cached and non-cached nodes and their definition hashes.
 
         Parameters
         ----------
-        observation_table_id: PydanticObjectId
-            Observation table id
+        feature_table_cache_metadata: FeatureTableCacheMetadataModel
+            Feature table cache metadata
+        graph: QueryGraph
+            Graph definition
+        nodes: List[Node]
+            Input node names
 
         Returns
         -------
-        Optional[FeatureTableCacheModel]
-            Feature Table Cache model if exists
+        Tuple[Dict[str, str], Dict[str, str]]
+            Tuple of
+            - cached definition hases to nodes mapping
+            - nont cached definition hases to nodes mapping
         """
-        documents = []
+        node_names = [node.name for node in nodes]
+        pruned_graph, node_name_map = graph.quick_prune(target_node_names=node_names)
```

Comment:
@gee-senbong Is this the right way of pruning the graph before extracting feature definition hashes, combined with the `prepare_graph_to_store` steps below?

Reply:
Opps, I missed this comment. The current flow seems correct to me.

Reply:
Just to clarify the logic here:

1. Quick prune here is to remove all used nodes from the input graph.
2. The following statement is used to prune the graph & clean up the view graph node metadata (same flow used to generate graph for feature definition, https://github.com/featurebyte/featurebyte/blob/63e1f00fddf4a426c059e24ec5009d0ff95aa538/featurebyte/worker/util/batch_feature_creator.py#L227-L230).
```
            (
                prepared_graph,
                prepared_node_name,
            ) = await self.namespace_handler.prepare_graph_to_store(
                graph=pruned_graph,
                node=pruned_graph.get_node_by_name(node_name_map[node.name]),
                sanitize_for_definition=True,
            )
```

Since the graph used for feature definition should the consistent with the graph that is saved in persistent (otherwise, feature definition should fail and user can't save the feature), it should be valid that is the final version of the feature's graph.